### DISCUSSION TITLE
Get-PnPFile Fails When Called From Azure Runbook

### DISCUSSION THREAD

**Discussion by dremillard** [pnp/powershell#2585] [Q&A]
Hi folks,  

I'm trying to download a file from SharePoint Online using **Get-PnPFile**.  I'm using an Azure App Registration and Cert Thumbprint to connect.  It works fine when I do it from my laptop but fails with a **"file not found 404"** when I do it from a PowerShell runbook in Azure.  

I'm running **PowerShell 7** and **PnP.PowerShell version 1.12**.   (Other things break if I use PowerShell 5).

Connect statement is like:
**$Conn = Connect-PnPOnline -Url $SiteURL -ClientId $ClientID -Tenant $TenantID -Thumbprint $Thumbprint -ReturnConnection**

Get-PnPFile statement is like:
**Get-PnPFile -Url $SiteTemplateURL -Path c:\temp -FileName $SiteTemplateFile -AsFile -Connection $Conn**

From examples I've seen online, $SiteTemplateURL seems to need a server relative address like "/sites/sitename/libraryname/documentname.xml".  I've also tried just "/libraryname/documentname.xml" to no avail. 

I've also tried putting a variable in front like:
**$status = Get-PnPFile -Url $SiteTemplateURL -Path c:\temp -FileName $SiteTemplateFile -AsFile -Connection $Conn**

I've also tried something I found in a Tony Redmond article:
**$status = (Get-PnPFile -Url $SiteTemplateURL -Path c:\temp -FileName $SiteTemplateFile -AsFile -Connection $Conn | Out-Null)**

Any ideas on why it would work on my laptop but not in an Azure Runbook?   Anyone bump into this before?  

Thanks in advance, 
Dave

**Comment by dremillard** [comment]
Surprised no one has encountered this fairly common scenario.  I would have thought it should work in an Azure Runbook, so it seemed like a bug.  It would be interesting to know if anyone has gotten it to work.   Anyone?

**Reply by AndersRask** [reply]
Not sure you are allowed to write to c:\temp in an automation account.

When I do similar stuff, I write to $env:temp

**Comment by Amitkumar028** [comment]
Use PnP.PowerShell version 1.7.0 with PowerShell 7.2 and Az runtime version to ~4, Hopefully it will work, latest PnP version are unstable with Az runbook

**Reply by dremillard** [reply]
Thanks, Amit.  Unfortunately, it's still giving the same error.  It seems that runbook execution isn't regression tested and not considered a priority in the PnP community.  IMHO, that's a big gap.   If I were a developer, I might be able to help, but unfortunately that's not the case.  It would be interesting to know if ANYONE has gotten an app security principal to work in an Azure runbook with PowerShell 7.2 and PnP.PowerShell 1.12.   Anyone at all?

**Reply by myatix** [reply]
I have the same issue in an Azure Function that I have created...

```
using namespace System.Net

param($Request)

$env:PNPPOWERSHELL_UPDATECHECK="false"
$tenant = $env:Tenant
$clientId = $env:ClientId
$certificateBase64Encoded = $env:Certificate
$templateLibUrl = $env:TemplateLibUrl
$templateFileName = $env:TemplateFileName
$downloadFilePath = $PSScriptRoot
$webUrl = $Request.Body.webUrl


$templateSiteUrl = ($templateLibUrl -Split '/')[0..4] -Join '/'
    write-host $templateSiteUrl
    $tplSiteConnect = Connect-PnPOnline -Url $templateSiteUrl -Tenant $tenant -ClientId $clientId -CertificateBase64Encoded $certificateBase64Encoded

$templateRelUrl = ($templateLibUrl -Split '/' | Select-Object -Skip 3) -Join '/'
    $templateFileUrl = "/" + $templateRelUrl + "/" + $templateFileName

    write-host "TemplateFileURL" $templateFileUrl
    write-host "downloadFilePath" $downloadFilePath
    write-host "TemplateFileName" $templateFileName
    Get-PnPFile -Url $templateFileUrl -Path $downloadFilePath -FileName $templateFileName -AsFile -Force -Connection $tplSiteConnect
    
    $templateFilePath = $downloadFilePath + "\" + $templateFileName
    Invoke-PnPSiteTemplate -Path $templateFilePath -Parameters @{"SiteUrl"=$webUrl} -Connection $webConnect
```

Error:

[Information]   INFORMATION: https://xxx.sharepoint.com/sites/SPOSiteProvisioning
2022-12-16T17:22:30Z   [Information]   INFORMATION: TemplateFileURL: /sites/SPOSiteProvisioning/Shared%20Documents/site-template.xml
2022-12-16T17:22:30Z   [Information]   INFORMATION: downloadFilePath: C:\home\site\wwwroot\SPOSiteProvisioning
2022-12-16T17:22:30Z   [Information]   INFORMATION: TemplateFileName: site-template.xml
2022-12-16T17:22:31Z   [Error]   ERROR: File Not Found.

Exception             : 
    Type                          : Microsoft.SharePoint.Client.ServerException
    ServerErrorCode               : -2147024894
    ServerErrorTypeName           : System.IO.FileNotFoundException
    ServerErrorTraceCorrelationId : 25e282a0-a00b-5000-c89c-96f9561fbb90
    TargetSite                    : 
        Name          : ProcessResponseStream
        DeclaringType : Microsoft.SharePoint.Client.ClientRequest
        MemberType    : Method
        Module        : Microsoft.SharePoint.Client.Runtime.dll
    StackTrace                    : 
   at Microsoft.SharePoint.Client.ClientRequest.ProcessResponseStream(Stream responseStream)
   at Microsoft.SharePoint.Client.ClientRequest.ProcessResponse()
   at Microsoft.SharePoint.Client.ClientRequest.ExecuteQueryToServerAsync(ChunkStringBuilder sb)
   at Microsoft.SharePoint.Client.ClientRequest.ExecuteQueryAsync()
   at Microsoft.SharePoint.Client.ClientRuntimeContext.ExecuteQueryAsync()
   at Microsoft.SharePoint.Client.ClientContext.ExecuteQueryAsync()
   at Microsoft.SharePoint.Client.ClientContextExtensions.ExecuteQueryImplementation(ClientRuntimeContext clientContext, Int32 retryCount, String userAgent)
   at Microsoft.SharePoint.Client.ClientContextExtensions.ExecuteQueryRetry(ClientRuntimeContext clientContext, Int32 retryCount, String userAgent)
   at PnP.PowerShell.Commands.Files.GetFile.SaveFileToLocal(Web web, String serverRelativeUrl, String localPath, String localFileName, Func`2 fileExistsCallBack) in c:\PnPPowerShell\src\Commands\Files\GetFile.cs:line 136
   at PnP.PowerShell.Commands.Files.GetFile.ExecuteCmdlet() in c:\PnPPowerShell\src\Commands\Files\GetFile.cs:line 82
   at PnP.PowerShell.Commands.PnPSharePointCmdlet.ProcessRecord() in c:\PnPPowerShell\src\Commands\Base\PnPSharePointCmdlet.cs:line 60
    Message                       : File Not Found.
    Data                          : System.Collections.ListDictionaryInternal
    Source                        : Microsoft.SharePoint.Client.Runtime
    HResult                       : -2146233088
CategoryInfo          : WriteError: (:) [Get-PnPFile], ServerException
FullyQualifiedErrorId : EXCEPTION,PnP.PowerShell.Commands.Files.GetFile
ErrorDetails          : File Not Found.
InvocationInfo        : 
    MyCommand        : Get-PnPFile
    ScriptLineNumber : 51
    OffsetInLine     : 5
    HistoryId        : 1
    ScriptName       : C:\home\site\wwwroot\SPOSiteProvisioning\run.ps1
    Line             : Get-PnPFile -Url $templateFileUrl -Path $downloadFilePath -FileName $templateFileName -AsFile -Force -Connection $tplSiteConnect
                       
    PositionMessage  : At C:\home\site\wwwroot\SPOSiteProvisioning\run.ps1:51 char:5
                       +     Get-PnPFile -Url $templateFileUrl -Path $downloadFilePath -FileNa …
                       +     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    PSScriptRoot     : C:\home\site\wwwroot\SPOSiteProvisioning
    PSCommandPath    : C:\home\site\wwwroot\SPOSiteProvisioning\run.ps1
    InvocationName   : Get-PnPFile
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, C:\home\site\wwwroot\SPOSiteProvisioning\run.ps1: line 51
PipelineIterationInfo :
